File: checks/ruby/lib/section508/checks/label_in_name.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Visible text of the control must appear in the accessible name.
    class LabelInName
      def self.call(root)
        return [] unless root&.respond_to?(:css)

        Dom.elements(root, "button, a[href], input, select, textarea").each do |control|
          visible = visible_text(control, root)
          next if visible.empty?

          name = accessible_name(control, root)
          unless name.downcase.include?(visible.downcase)
            return ["The visible text “#{visible}” is not in the accessible name “#{name}”."]
          end
        end
        []
      end

      def self.visible_text(control, root)
        if %w[BUTTON A].include?(control.name.upcase)
          return Dom.norm(control)
        end

        id = control["id"]
        if id && !id.empty?
          label = Dom.label_for(root, id)
          return Dom.norm(label) if label
        end

        wrap = Dom.closest(control, "label")
        return "" unless wrap

        Dom.label_text(wrap, control)
      end

      def self.accessible_name(control, root)
        if Dom.has_attr?(control, "aria-label")
          return control["aria-label"].to_s.gsub(/\s+/, " ").strip
        end

        labelledby = control["aria-labelledby"]
        if labelledby && !labelledby.empty?
          text = labelledby.split(/\s+/).map { |id| node = Dom.by_id(root, id); node ? Dom.text(node) : "" }.join(" ")
          return text.gsub(/\s+/, " ").strip
        end

        return Dom.norm(control) if %w[BUTTON A].include?(control.name.upcase)

        visible_text(control, root)
      end
    end
  end
end

File: checks/python/src/section508/checks/label_in_name.py
from section508.dom import (
    SPACE,
    attr,
    clone,
    closest,
    find_id,
    has_attr,
    label_for,
    remove_matches,
    select,
    tag_name,
)


def check(root):
    if not root:
        return []

    for control in select(root, "button, a[href], input, select, textarea"):
        visible = visible_text(control, root)
        if not visible:
            continue
        name = accessible_name(control, root)
        if visible.lower() not in name.lower():
            return [f"The visible text “{visible}” is not in the accessible name “{name}”."]
    return []


def visible_text(control, root) -> str:
    if tag_name(control) in {"button", "a"}:
        return SPACE.sub(" ", control.get_text()).strip()
    control_id = attr(control, "id")
    if control_id:
        label = label_for(root, control_id)
        if label:
            return SPACE.sub(" ", label.get_text()).strip()
    wrap = closest(control, "label")
    if not wrap:
        return ""
    copied = clone(wrap)
    remove_matches(copied, "input, select, textarea, button")
    return SPACE.sub(" ", copied.get_text()).strip()


def accessible_name(control, root) -> str:
    if has_attr(control, "aria-label"):
        return SPACE.sub(" ", attr(control, "aria-label") or "").strip()
    labelledby = attr(control, "aria-labelledby")
    if labelledby:
        parts = []
        for ref in labelledby.split():
            node = find_id(root, ref)
            parts.append(node.get_text() if node is not None else "")
        return SPACE.sub(" ", " ".join(parts)).strip()
    if tag_name(control) in {"button", "a"}:
        return SPACE.sub(" ", control.get_text()).strip()
    return visible_text(control, root)
